source/developer/redux.rst

```diff
@@ -0,0 +1,261 @@
+
+
+Adding a Selector
+------------------
+
+Selectors are the method used to retrieving data from the state of the store. We use `reselect <https://github.com/reactjs/reselect>`__. If you'd like to know more about reselect and how we use it at Mattermost, `check out this developer talk given by core developer Harrison Healey <https://www.youtube.com/watch?v=6N2X7gEwmaQ>`__.
+
+Selectors must:
+- Receive ``state`` as the first argument and data returned must be solely based on what's in the state
+- Be created with ``createSelector`` whenever the data is manipulated or formatted before return
+- Be unit tested
+
+Selectors live in the ``src/selectors/`` directory.
+
+Implementing the Selector
+~~~~~~~~~~~~~~~~~~
+
+If your selector is just pulling data directly from the state without any manipulation, simply just return the data you need.
+
+
+.. code-block:: javascript
+  :linenos:
+
+  export function getUser(state, id) {
+      return state.entities.users.profiles[id];
+  }
+
+The above example is just simply pulling a user out of the profiles entity and requires no computation or formatting.
+
+If your selector needs to select based on some more advanced requirements or needs the result in a specific format then you'll need to make use of the ``createSelector`` function from `reselect <reselect <https://github.com/reactjs/reselect>`__. If you're not sure what this is good for, `check out the previously mentioned developer talk <https://www.youtube.com/watch?v=6N2X7gEwmaQ>`__. The short form reason is using reselect allows for memoization and only runs the computation of selectors when the state affecting that selector has actually changed.
+
+The basic usage for ``createSelector`` is pass it all the selector functions needed as inputs to your computation. The last argument will then be a function that takes in the results of each selector you specified and performs your computation on that data before returning it.
+
+.. code-block:: javascript
+  :linenos:
+
+  export const getUsersByUsername = createSelector(
+      getUsers,
+      (users) => {
+          const usersByUsername = {};
+
+          for (const id in users) {
+              if (users.hasOwnProperty(id)) {
+                  const user = users[id];
+                  usersByUsername[user.username] = user;
+              }
+          }
+
+          return usersByUsername;
+      }
+  );
+
+Here we're using the ``getUsers`` selector to feed users into our function that builds a map of users with username as the key.
+
+So far that's pretty straight forward but what if you want to select some data based on an argument but also need to do some computation? That is a little more tricky if you haven't wrapped your head around the purpose of reselect and how createSelector works, so if you haven't watched the developer talk linked above I would strongly suggest it.
```

Comment:
The first sentence is a bit awkward with the double "but"s. Maybe remove the second one and the bit following it like:
"So far that's pretty straightforward, but what if you want to select some data based on an argument?"

"straight forward but what if" -> "straightforward, but what if"
"linked above I would" -> "linked above, I would"